telegram/ext/builders.py

```diff
@@ -0,0 +1,1338 @@
+        # * for each of the workers
+        # * 1 for Dispatcher
+        # * 1 for Updater (even if webhook is used, we can spare a connection)
+        # * 1 for JobQueue
+        # * 1 for main thread
+        con_pool_size = self._workers + 4
+        actual_size = dispatcher.bot.request.con_pool_size
+        if actual_size < con_pool_size:
+            warnings.warn(
+                f'The Connection pool of Request object is smaller ({actual_size}) than the '
+                f'recommended value of {con_pool_size}.',
+                stacklevel=2,
+            )
+
+        return dispatcher
+
+    def _build_updater(
+        self: '_BaseBuilder[ODT, BT, Any, Any, Any, Any, Any, Any]',
+    ) -> Updater[BT, ODT]:
+        if self._dispatcher_was_set is False:
+            dispatcher = self._build_dispatcher()
+            return self._updater_class(
+                dispatcher=dispatcher,
+                user_signal_handler=self._user_signal_handler,
+                exception_event=dispatcher.exception_event,
+                builder_flag=True,
+                **self._updater_kwargs,
+            )
+        return self._updater_class(
+            dispatcher=self._dispatcher,
+            bot=self._dispatcher.bot if self._dispatcher else (self._bot or self._build_ext_bot()),
+            update_queue=self._update_queue,
+            user_signal_handler=self._user_signal_handler,
+            exception_event=self._dispatcher.exception_event
+            if self._dispatcher
+            else self._exception_event,
+            builder_flag=True,
+            **self._updater_kwargs,
+        )
+
+    @staticmethod
+    def _exception_builder(arg_1: str, arg_2: str = None) -> RuntimeError:
+        if not arg_2:
+            return RuntimeError(f'The parameter `{arg_1}` was already set.')
+        return RuntimeError(f'The parameter `{arg_1}` can only be set, if the no {arg_2} was set.')
+
+    @property
+    def _dispatcher_check(self) -> bool:
+        return self._dispatcher_was_set and self._dispatcher is not None
+
+    @check_if_already_set
+    def _set_dispatcher_class(
+        self: BuilderType, dispatcher_class: Type[Dispatcher], kwargs: Dict[str, object] = None
+    ) -> BuilderType:
+        if self._dispatcher_was_set:
+            raise self._exception_builder('dispatcher_class', 'Dispatcher instance')
+        self._dispatcher_class = dispatcher_class
+        self._dispatcher_kwargs = kwargs or {}
+        self._dispatcher_class_was_set = True
+        return self
+
+    @check_if_already_set
+    def _set_updater_class(
+        self: BuilderType, updater_class: Type[Updater], kwargs: Dict[str, object] = None
+    ) -> BuilderType:
+        self._updater_class = updater_class
+        self._updater_kwargs = kwargs or {}
+        self._updater_class_was_set = True
+        return self
+
+    @check_if_already_set
+    def _set_token(self: BuilderType, token: str) -> BuilderType:
+        if self._bot_was_set:
```

Comment:
Should we assert that `self._request` wasn't set as well?

Reply:
not really. both `token` and `request` can bes passed to `Bot`, so if `_request` was set you can still set a token …